Feature/soft build

juewu4072 opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->
This is a commit about the compilation process.
1. Add submodule, and introduce agent-libs as git-submodules to improve locally compilation efficiency.
2. Add make command line to facilitate compilation.

## Related Issue
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an open issue describing it with details -->
<!--- Please link to the issue here: -->

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
Each time you compile and include the probe locally, you need to extract agent-libs from github twice. The version mapping between agent-libs and kindling is not as clear as the mapping between git submodules. After the git submodule is introduced, git will record the version correspondence. The compilation process contains too many compile commands, which is prone to errors. It is best to use makefiles directly

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
use command to build
```
$cd kindling/
$make docker-build-all
```
After successful compilation
```
$docker images
```

![image](https://user-images.githubusercontent.com/37373487/216323060-afe7fa45-dfa5-498c-85d0-a6f8636d0384.png)

sanyangji commented:
Good Idea.
I follow your test but find some problems in Makefile. I fail to build the image, can you do more tests and fix them? 
Besides, some new updates should be merged into Dockerfile and take care of typos and code indentation.
One more thing, I'm not sure whether you should take compiling kindling-falcolib-probe tarball into account, as in Makefile, `make probe-libs first`, which will make the compilation process complete.

juewu4072 commented:
sorry to change that bug, the wrong name for the build-image has been fix, and the docker-build-all command was retested successfully.

The dockerfile was modified because it took too much time to install debugging tools such as gdb when packaging the image. we can give them more choices, such as the two common scenarios supported:
1. Need debug tools, need local probe 
2. No debugging tools required, need compile probes

`make probe-libs first` is a good idea, it works well when my computer kernel version is the same, I also use it, since my friend and I compiled with source code on our own computer, because our kernel-version is different, we need to execute make probe-libs every time, to simplify this process, I added several different commands in the makefile for users to choose, for example, 
1. `make dependencies` to install deps
2. `make agent-libs` to build probe-libs
3. `make kindling-collector` to  build collector image
4. `make docker-build-all` to install deps and build probe-libs and build cooector image
 
I added new comments to makefile. @sanyangji

sanyangji commented on Makefile:
```diff
@@ -13,4 +13,135 @@ collector: libkindling
+
+.PHONY: build-collector
+build-collector:
+	@cd $(SCRIPTS_PATH) && sh ./build.sh Dokcerfile
```
typo: Dockerfile

sanyangji commented on collector/docker/Dockerfile:
```diff
@@ -1,14 +1,13 @@
 FROM centos:7
 
 WORKDIR /app/
-RUN yum install -y qt5-qtbase-devel
-RUN yum install -y gdb
-COPY gdb_print.sh /app/
```
these are used to load kindling plugin and to export information when unexpected exit happens https://github.com/KindlingProject/kindling/pull/398

sanyangji commented on Makefile:
```diff
@@ -13,4 +13,135 @@ collector: libkindling
+  		echo "$(LIBS_SRC) not exist, download....."; \
+		git submodule update --init --recursive; \
+	else \
+	  echo "agent-libs: $(LIBS_SRC) exist"; \
```
inconsistent code indent

sanyangji commented on Makefile:
```diff
@@ -13,4 +13,135 @@ collector: libkindling
+	@if [ ! -d "$(LIBS_SRC)/kindling-falcolib-probe/" ]; then \
+    		echo "The packaged probe does not exist.try again"; \
+    		exit 1; \
+    else \
```
inconsistent

sanyangji commented:
Sorry, I misunderstood the meaning of `agent-libs`. `agent-libs` is composed of userspace part (static libs, you may call libs or agent-libs) and kernel space part (that is kindling-falcolib-probe, you may call it  kindling-falcolib-probe or probe but without libs). So can you rename them to make them clear? Thanks to your comments anyway.

Some comments have been added and I'm still trying to test, please wait a moment.

juewu4072 commented:
I've fixed typos, bad naming, code indentation.

my idea has some problems with indent and tab  T_T

dxsup commented on .gitmodules:
```diff
@@ -0,0 +1,3 @@
+[submodule "agent-libs"]
+	path = agent-libs
```
Since `agent-libs` is one of the dependencies of `probe`, I think we better put the submodule under `kindling/probe`. What do you think?

juewu4072 replied:
The agent library is a dependency of the probe, and I think it makes sense to do that. I'll put it under the kindling/probe/libs/.

dxsup commented on .gitmodules:
```diff
@@ -0,0 +1,3 @@
+[submodule "agent-libs"]
+	path = agent-libs
+	url = https://github.com/KindlingProject/agent-libs.git
```
I suggest we write down the branch name explicitly to make things clear.

juewu4072 replied:
Branching of submodules is allowed, but this is inconsistent with the original design of the submodules, and submodules with specific commits can be used with the current code. Same as libbpf/libbpf bootstrap

juewu4072 replied:
`https://github.com/libbpf/libbpf-bootstrap/blob/master/.gitmodules`

dxsup replied:
`git submodule` does record a specific commit to allow users to pull the correct repository version. There is no question that the current `.gitmodules` works. AFAIK, tracking a remote branch can make things clear because we can use `git submodule update --remote` to update the code. See [How can I specify a branch/tag when adding a Git submodule?](https://stackoverflow.com/questions/1777854/how-can-i-specify-a-branch-tag-when-adding-a-git-submodule).

It is OK if we use the current `.gitmodules` and make an improvement later.

juewu4072 replied:
Thank you for your help. That's a great idea. Adding submodule branches is a great way to improve multiplayer collaboration, and it solves a problem I had with using submodule multiplayer collaboration scenarios. Thanks again

juewu4072 replied:
agent-libs has moved to 'kindling/probe/libs/agent-libs' and add branch mark in .gitmodules

dxsup commented on deploy/scripts/build.sh:
```diff
@@ -1,22 +1,33 @@
+  mkdir -p ../../collector/docker/libso && cp -rf ./src/libkindling.so ../../collector/docker/libso/
   cp -rf ./src/libkindling.so /usr/lib64/
   cd ../../collector/
+  go mod tidy && go mod vendor
```
Is `go mod vendor` necessary?

juewu4072 replied:
Has been removed in the current version, this will slow down the image packaging speed

dxsup replied:
AFAIK, `go mod vendor` is helpful in two cases:
1. When there is no Internet access.
2. When some packages are not available from the proxy anymore.

Kindling has been using `GOPATH` as the pkg directory and it works well. I have not seen the need to use `go mod vendor`.

Regarding the image packaging speed, unless we maintain the `vendor` directory in the Git repository, there is no noticeable difference, IMO.

juewu4072 replied:
When we compile  `GOPATH = /root/go/` ，use run_docker. sh again, path ‘/root/go’ does not exist, need to download modules again. If we do vendors, the vendor path  `/kindling/collector/vendor/` will be mapped to images with Kindling code. Compiling again will reduce the time. This is only valid for multiple partial compilations. 

It's a great help to compile multiple times when we're local.

juewu4072 replied:
This scenario is friendly for local build testing, but has some concerns about being a generic scenario, so I've removed it in a new commit, which should be done by tomorrow evening at the latest

juewu4072 replied:
removed 'go mod vendor'

dxsup commented:
Are you still working on this PR? Hope for your response.

dxsup commented:
Thanks for your work again. I'll have a test after which I'll merge this PR.

dxsup commented:
@sanyangji @jundizhou Do we still need to maintain the commit hash of `agent-libs` in `CMakeLists.txt` since this PR has introduced `git submodule`?

sanyangji commented:
> @sanyangji @jundizhou Do we still need to maintain the commit hash of `agent-libs` in `CMakeLists.txt` since this PR has introduced `git submodule`?
I have no idea about the usage of the `commit` if merging this PR. It's better to remove it.

NeJan2020 commented on Makefile:
```diff
@@ -13,4 +13,144 @@ collector: libkindling
+		exit 1; \
+	else \
+		cd $(LIBS_SRC) && \
+		tar -cvzf kindling-falcolib-probe.tar.gz kindling-falcolib-probe/ && \
```
Please add the `probe/libs/agent-libs/kindling-falcolib-probe.tar.gz` to the .gitignore of `agent-lib`, or just remove it after copy.

juewu4072 replied:
Changed `cp -f` to `mv`

juewu4072 replied:
Have also added `clean` to the makfile, clean the build file

juewu4072 commented:
> 

Removed useless patch

dxsup commented on Makefile:
```diff
@@ -13,4 +13,149 @@ collector: libkindling
+.PHONY: patch-libs
+patch-libs:
+	@echo "Checking libs...";
+	@if [ ! -d "$(LIBS_SRC)" ]; then \
+		echo "$(LIBS_SRC) not exist, downloading....."; \
+		git submodule update --init --recursive; \
```
The `if` statement is always false because `agent-libs` is a directory and it's always there once you clone a fresh Kindling repository.

juewu4072 replied:
Yes, your findings are valuable, and that really shouldn't be judged by `if -d $path`.

juewu4072 replied:
Changed to `ls -A1q $path` to check dir is empty

dxsup commented on Makefile:
```diff
@@ -13,4 +13,149 @@ collector: libkindling
+kernelVersion = $(shell uname -r)
+ErrorExit = exit 1
+
+ifneq ($(shell cat /etc/os-release | grep Ubuntu | wc -l),0)
+	KernelName = linux-headers-$(kernelVersion)
+	CheckCommand = dpkg -l | grep $(KernelName)
+	InstallCommand = sudo apt-get -y install $(KernelName)
+else
+	ifneq ($(shell cat /etc/os-release | grep CentOS | wc -l),0)
+		KernelName = kernel-devel-$(kernelVersion)
+		CheckCommand = yum list installed |grep $(KernelName)
+		InstallCommand = sudo yum -y install $(KernelName)
+	else
+		KernelName = NotSupport
+		CheckCommand = $(ErrorExit)
+		InstallCommand = $(ErrorExit)
+	endif
+endif
```
Looks like these are only needed when running `make kernel-headers`. Why not move these under that part?

By the way, I figured `CheckCommand` is never used and should be removed.

juewu4072 replied:
Moved to `make kernel-headers`.

dxsup reviewed (approved):
LGTM

dxsup commented:
I'm writing down TODOs here once we merge this PR:
- [ ] Fix the action `Kindling-agent-CI`.
- [ ] Rewrite the doc guiding developers on how to build.
